Add Debian Packaging for Linux

usaleem-ix opened the pull request:
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Provide a general summary of your changes in the Title above -->

<!---
Documentation on ZFS Buildbot options can be found at
https://openzfs.github.io/openzfs-docs/Developer%20Resources/Buildbot%20Options.html
-->

### Motivation and Context
Add Debian Packaging for Linux

### Description
<!--- Describe your changes in detail -->
Currently, the Debian packages are generated from ALIEN that converts
RPMs to Debian packages. This commit adds native Debian packaging for
Debian based systems from ZFS on Linux.

Proper Debian style packaging can:

1. Make things easier for developers to have an actively managed packaging for Debian based systems.
2. Ease the testing and development of ZFS packages on Debian based systems.
3. It may also serve as a starting point for downstream projects to base their packaging on and add their own customization.
4. Current .deb packages do not include debug symbols and information to facilitate debugging process.

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->
<!--- Please think about using the draft PR feature if appropriate -->
Built the packages from new packaging. Installed the packages on Debian 11 and ran the test suite.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

gdevenyi commented:
Can you please add a complimentary PR to https://github.com/openzfs/openzfs-docs to update the docs for debian based systems?

usaleem-ix commented:
> Can you please add a complimentary PR to https://github.com/openzfs/openzfs-docs to update the docs for debian based systems?

@gdevenyi Sure, will do.

rincebrain commented:
Quite a long way to just fix debuginfo generation, heh. Well done.

My only suggestion is that it might be good to acknowledge that this is a minimally modified copy of the existing Debian maintainers' packaging for OpenZFS somewhere other than debian/copyright? I don't know that it's _necessary_, but I had to go looking to confirm you hadn't just deleted all references to them, since you emptied the changelog, and I'd rather acknowledge people too many times than too few. :)

gmelikov commented on debian/copyright:
```diff
@@ -0,0 +1,1278 @@
+Format: https://www.debian.org/doc/packaging-manuals/copyright-format/1.0/
```
If this file is necessary - could we generate it, not hard-push it and forget to update?

usaleem-ix replied:
Yes, this file can generated using some tools as highlighted [here](https://wiki.debian.org/CopyrightReviewTools). But that requires a bit more work than manually updating the file in my opinion. If you recommend, I can remove this file.

cdluminate replied:
copyright claims for `debian/*` scripts are contained in this file. Please don't remove it ......

behlendorf commented:
@happyaron @xnox @aerusso @cdluminate would you mind commenting on this?  We'd like to (finally) add proper debian style packaging to the upstream repository to make it easier to build real packages to test with.  However, I want to make sure we avoid stepping on any toes or causing unnecessary problems downstream.  What's the right way for us to go about this?

happyaron commented:
> 

One of the possible scenario I can think of is people who run a custom build of debian package would like to upgrade to a distro build at a later time, or vice versa. Some care of the package names and dependencies should be taken to make sure they are compatible with the distro ones, so that a valid upgrade path exists.

pzakha commented:
I'm curious, did you use existing Ubuntu ZFS packages as reference?

cdluminate commented:
Ubuntu ZFS package is pulled from Debian with slight modification. I'm continuously synchronizing Ubuntu modifications back to Debian package. But in fact most of the changes made by Ubuntu are Ubuntu-specific, like zsys support. They are not enabled on Debian.

cdluminate commented:
I currently have several concerns regarding this:
(1) is it possible to not place the `debian` directory in root directory? That conflicts with our workflow and will force us to modify the upstream source tarball.
(2) how do we synchronize changes? 
(3) are your sure the `2.1.99` version number in debian/changelog is intended? Namely, a locally built, say 2.1.2 in fact will always suppress debian's 2.1.X release because `dpkg --compare-versions 2.1.99 gt 2.1.X` is always true.

nabijaczleweli commented:
Yes, and I consider putting packaging in upstream sources idiotic at best, *especially* Debian-style packaging that takes over the source directory, *especially especially* if the downstream heavily patches this and has a particular binary package layout. If it were up to me I'd close this as severely out-of-scope; but it isn't, so I didn't post here at all.

But (a) if you put `debian/` in the root then Debian will have to repack, so (b) it's best to move rpm -> contrib/rpm and put this in contrib/debian (or package/ if the Laboratory actively supports the RPM packaging), but then (c) it won't work OOB, obviously, so (d) *don't* inline Debian packaging.

usaleem-ix commented:
> I'm curious, did you use existing Ubuntu ZFS packages as reference?

@pzakha I used existing ZFS on Linux packaging for Debian, that is available [here](https://salsa.debian.org/zfsonlinux-team/zfs).

usaleem-ix commented:
> I currently have several concerns regarding this: (1) is it possible to not place the `debian` directory in root directory? That conflicts with our workflow and will force us to modify the upstream source tarball. (2) how do we synchronize changes? (3) are your sure the `2.1.99` version number in debian/changelog is intended? Namely, a locally built, say 2.1.2 in fact will always suppress debian's 2.1.X release because `dpkg --compare-versions 2.1.99 gt 2.1.X` is always true.

@cdluminate 

1. I think Debian packaging requires that `debian` directory should be placed on the root of the source directory. If we want to change the location of this directory and place it on root afterwards when we would want to build.
2. We would either be required to synchronize this as is, or we can take it from this point and continue to maintain this.
3. Actually I had used `2.1.X` as version number in debian/changelog but I changed it to `2.1.99` to make it consistent with [META](https://github.com/openzfs/zfs/blob/master/META). If you recommend, I can change it back to `2.1.X`.

cdluminate commented:
(2) I can't agree with this as a good engineering practice. Assume Debian's packaging commit A has a bug. This snapshot is also imported to openzfs upstream. Is this bug debian bug or upstream bug?
As time goes by, we debian fixed that bug and commited as B. But openzfs/zfs still use debian packaging commit A. Is this bug debian bug or upstream bug then? Debian packaging bug DO EXIST.

usaleem-ix commented:
We do agree that this is not a good approach. Moving `debian`  to `contrib` and rename the packages to `openzfs-*` can help clarify that this is not expected to be Debian `zfsonlinux` package set. I am working on that front. This can hopefully make things easier for many developers to have an actively maintained packaging for Debian based systems.

For synchronizing changes, any changes downstream that seem generally useful may be contributed back to upstream since git allows to merge into a sub-directory.

cdluminate commented:
> We do agree that this is not a good approach. Moving `debian` to `contrib` and rename the packages to `openzfs-*` can help clarify that this is not expected to be Debian `zfsonlinux` package set. I am working on that front. This can hopefully make things easier for many developers to have an actively maintained packaging for Debian based systems.
> 
> For synchronizing changes, any changes downstream that seem generally useful may be contributed back to upstream since git allows to merge into a sub-directory.

In that case please specify Conflicts+Replaces in `debian/control` to make sure locally-packaged `openzfs-*` packages cannot coinstall with `zfs-*`. We will have to add that from Debian side as well. https://www.debian.org/doc/debian-policy/ch-relationships.html

Fabian-Gruenbichler commented:
as a former occasional Debian zfs contributor, and one of the downstream maintainers of [ZFS packages derived from Debian's](https://git.proxmox.com/?p=zfsonlinux.git;a=summary) - another option instead of fully renaming the packages is to use a version suffix but keep the names/contents broadly in sync. we do this in our downstream distros (which are using the main Debian repositories and a downstream specific one, with the latter shipping higher versions of packages also shipped by Debian). Debian backports does the inverse (using a suffix that sorts lower than the actual package, so that upgrading from N-backports to N+1 works), Ubuntu uses a similar scheme for some packages.

obviously, this requires some amount of coordination but not that much:
- patches cherry-picked for compat/bug fix reasons in Debian obviously shouldn't be included upstream, but patches which exist to smooth building as part of the package should
- conflicts arising from upstream changes can be fixed in upstream packaging and Debian notified (since they likely need to be included when the upstream version including these changes is included in Debian)
- bugs in packaging fixed in Debian should be pushed upstream as well (either to the release branch only or both master and release branches, depending on where they are relevant)

IMHO the most important requirement is that there needs to be someone available for reviewing changes to the upstream/forked packaging that (really) understands how Debian packaging works, and has a good working relationship with downstream (Debian/Ubuntu) maintainers. not sure whether such a person exists or whether any of the existing downstream maintainers want/have capacity to fill that role.

dpkg-buildpackage operates on unpacked source trees, so if `debian/` is not included directly in the source tree some sort of wrangling needs to be employed by the Makefile (e.g., copying the `debian/` dir to the top level as part of building the packages), but that doesn't hurt that much and is probably preferable to annoying downstream packagers ;)

cdluminate commented:
Let me add my disclaimer here: As an active downstream Debian ZFS package maintainer, I will not keep track of the status and bugs of this copy of imported debian packaging. When new commits are made on the Debian side, I'll not inform the upstream to update the code copy. As long as (1) packaging incorporation does not interfere with the downstream (Debian) workflow, (2) copyright, changelog, credit to contributors and license are intact, I'm not going to stand in the way of anyone.

cdluminate commented:
Another detail -- the debian packaging is licensed under GPL-2+ due to historical reasons: https://salsa.debian.org/zfsonlinux-team/zfs/-/blob/master/debian/copyright#L194-204 -- please be aware.

xnox commented:
If motivation is described a bit better, different solutions and approaches can be done to make things easy to achieve the desired goal and make it easy to share packaging.

Are you planning to start providing snapshot / daily builds of openzfs for debian/ubuntu? Or is it an advanced-user / developer enhancement to allow one to self-build .debs and install them to see how things are interacting?

From Ubuntu point of view, it would be nice to use different package names, and maybe even have extra conflicts. This is because systems that are installed with zsys feature is currently trivially to irreversably make unbootable, and loose automatic snapshots and rollbacks. Thus ideally the packaging upstream should intentionally use different package names. Also in Ubuntu, zfs.ko is signed to support secureboot with enforced module signature checking. Care should be taken to ensure that packages provided here, do not allow installing dkms based zfs.ko module without signatures from an enrolled MOK key.

usaleem-ix commented:
> If motivation is described a bit better, different solutions and approaches can be done to make things easy to achieve the desired goal and make it easy to share packaging.
> 
> Are you planning to start providing snapshot / daily builds of openzfs for debian/ubuntu? Or is it an advanced-user / developer enhancement to allow one to self-build .debs and install them to see how things are interacting?
> 
> From Ubuntu point of view, it would be nice to use different package names, and maybe even have extra conflicts. This is because systems that are installed with zsys feature is currently trivially to irreversably make unbootable, and loose automatic snapshots and rollbacks. Thus ideally the packaging upstream should intentionally use different package names. Also in Ubuntu, zfs.ko is signed to support secureboot with enforced module signature checking. Care should be taken to ensure that packages provided here, do not allow installing dkms based zfs.ko module without signatures from an enrolled MOK key.

The need for proper Debian style packaging was felt to: 

(1) Make things easier for developers to have an actively managed packaging for Debian based systems.
(2) Ease the testing and development of ZFS packages on Debian based systems.
(3) It may also serve as a starting point for downstream projects to base their packaging on and add their own customization.
(4) Current .deb packages for OpenZFS do not include debug symbols and information to facilitate debugging process.

Ease of development, testing and debugging for Debian based systems is the only goal here.

cdluminate commented:
As for (4). Debian packages built using the standard process do not include debug symbols neither. Debugging information will be split into `-dbgsym` packages. Why not try to prevent rpm/alient from stripping binaries at the first place? But prefer adding 3000+ lines of code?

usaleem-ix commented:
> As for (4). Debian packages built using the standard process do not include debug symbols neither. Debugging information will be split into `-dbgsym` packages. Why not try to prevent rpm/alient from stripping binaries at the first place? But prefer adding 3000+ lines of code?

I did take that route first. But after discussing it on the slack channel for OpenZFS, it was recommended we take this as a shorter path than trying to convince alien to emit debugging information. And that we should try to switch to a less alien based generation system.

cdluminate commented:
I see. Sounds good to me.

cdluminate commented:
My mind has changed a little bit. For serious debian packaging bugs I'll notify the upstream if this PR eventually gets merged and suggest fix. Minor ones do not worth that and you may cherry pick by yourselves.

rincebrain commented:
As the one who last submitted a patch when he found a bug in alien causing problems, +1 on finding a solution for generating deb packages that is not reliant on alien, be it actual native packaging format or (though it doesn't seem likely now) if we couldn't find a solution that wouldn't cause Debian's maintainers more work, I believe at least one person maintains their own alternate scripting for generating OpenZFS debs from specfiles that isn't in front of me at the moment...

Conan-Kudo commented:
> > As for (4). Debian packages built using the standard process do not include debug symbols neither. Debugging information will be split into `-dbgsym` packages. Why not try to prevent rpm/alient from stripping binaries at the first place? But prefer adding 3000+ lines of code?
> 
> I did take that route first. But after discussing it on the slack channel for OpenZFS, it was recommended we take this as a shorter path than trying to convince alien to emit debugging information. And that we should try to switch to a less alien based generation system.

It's not a question of convincing alien to emit debugging information. You just have to tell rpmbuild to not strip the symbols before alien runs on the built packages. `--define '__strip /bin/true'` would have been enough.

Conan-Kudo commented:
That said, Debian generally prefers that upstream projects _not_ have a `debian` folder, because it's a pain for them to overwrite it when they have to mangle the source tree for their own packaging.

Additionally, I do have a way to build Debian packages from an RPM spec file natively, as [I've been doing this for years](https://pagure.io/openzfs-linux-packaging/blob/main/f/zfs.spec). I haven't contributed it upstream because I wasn't sure if people wanted it.

usaleem-ix commented:
Since including Debian packaging in upstream is not a good practice and it creates a over head for maintainers, I have decided not to pursue this path further.

Thanks to @Conan-Kudo, debug symbols can be included in existing packaging.

usaleem-ix commented:
> I wish you hadn't chosen this. It's not like it's a choice between "no Debian packages" and "Debian packages", it's a choice between "nice Debian packages" and "minimal Debian packages", after you already did all the work required and people had largely agreed.

Sure, I can reopen this PR once I have it ready with review comments addressed.

usaleem-ix commented:
I have incorporated the feedback given earlier:

- `debian` directory has been moved under `contrib`
- Package names have been modified to `openzfs-*`
- Conflicts have been added for standard Debian packages
- All copyright, changelog, credit to contributors has been kept intact

terem42 commented:
I wholeheartedly support you.
A couple of years ago I made [my own version](https://github.com/terem42/zfs-debian) of zfs packages for Debian 
I rebuild them off the master for new versions periodically, specifically to solve the problem described in this PR.
It's better to have it as part of an official build.

cdluminate commented:
I propose a potentially better solution for this Debian packaging issue. Let me know what you think about it.

I have seen many people periodically rebasing the Debian packaging to the ZFS master branch. Can we fork ZFS somewhere else and add Debian packaging there instead? A group of people (including me) can maintain the packaging branch collaboratively, and periodically rebase it against the ZFS master. In this way, the upstream does not need to merge the Debian packaging into the master branch, and inherit Debian-specific issues polluting the GitHub issues. And Git commits can be propagated into the Debian archive for ZFS releases, and the distro maintainers do not have to push updates everywhere upon bug fixes.

ZFS releases are already packaged for Linux distributions. So in my proposal, I only consider the packaging for the ZFS master branch.

freqlabs commented:
Packaging master for distros to ship in their repos is not the intent. Some developers find the ability to build packages convenient for working on master. The packaging may also be convenient for distros when it eventually is in a stable branch, but I fully expect different distros would largely have their own forks with whatever modifications, if not their own entirely separate packaging. It's not supposed to be some sort of official OpenZFS master package for Debian.

freqlabs commented:
@usaleem-ix do I understand correctly that all we have left to do is trim down the copyright file?

usaleem-ix commented:
@freqlabs yes trim down the copyright file, and rebase this on master to validate the latest bits.

cdluminate commented:
> Packaging master for distros to ship in their repos is not the intent. Some developers find the ability to build packages convenient for working on master. The packaging may also be convenient for distros when it eventually is in a stable branch, but I fully expect different distros would largely have their own forks with whatever modifications, if not their own entirely separate packaging. It's not supposed to be some sort of official OpenZFS master package for Debian.

There is ambiguity somewhere. I meant leaving the packaging files somewhere and refrain from merging them into the openzfs/zfs master branch, nor uploading to any linux distributions. It is exactly because some developers find it useful, that we leave the packaging files somewhere and keep it updated.

For stable releases of openzfs/zfs I don't think shipping a debian directory is useful in any sense. Debian keeps the original copy of the packaging file. Ubuntu pulls it to the downstream. Other ubuntu-based distributions pulls Ubuntu packages. Providing yet-another trimmed and different version of reference debian packaging file in openzfs/zfs master tree is not helpful for people who don't build debian package for openzfs/zfs master for their own.

I'm empirically the maintainer for the original copy of Debian packaging. In the case where this PR would be merged, I'll not submit PR here on the update of the original copy of debian packaging since it obviously doubles my work. The proposed packaging for openzfs/zfs avoids doubling my work so I'll be able to participate -- since it moves the original copy of Debian packaging to the upstream.

freqlabs commented:
We do development work on the stable branch as well, not just master.

For TrueNAS we are currently using the in-tree alien package build, because the official Debian packaging for zfs had not been updated for the major restructuring in OpenZFS 2.0 yet when we started putting together the build systems for SCALE. Having packaging integrated in this repo with CI means we don't have to wait for a downstream project to update their packages when the next stable branch is cut.

So it is useful on the stable branch, and it would be nice to not lose that.

cdluminate commented:
Ok, sounds sensible. Now I see that this is the best possible solution for you. Just keep in mind to explicitly preserve the names of the past contributors when trimming `debian/copyright`, and avoid putting the `debian` directory to the root directory of `openzfs/zfs`. In this way things should be fine for both sides. Ping me when ready for another round of review or technical suggestions on editing debian packaging files.

freqlabs commented on debian-packaging.sh:
```diff
@@ -0,0 +1,36 @@
+#!/bin/bash
```
This script should probably go in either contrib/debian/ or scripts/

usaleem-ix replied:
Moved to scripts/ dir.

freqlabs commented on contrib/debian/zfsutils-linux.zfs-share.init:
```diff
@@ -0,0 +1 @@
+../etc/init.d/zfs-share
```
Can we get rid of all this zfsutils-linux stuff? We don't build a zfsutils-linux package as far as I can tell.

freqlabs commented on contrib/debian/watch:
```diff
@@ -0,0 +1,2 @@
+version=4
+https://zfsonlinux.org/ .*zfs-([\d\.]+)\.tar\.gz$ debian uupdate
```
I don't think we need this either. It's for uscan to check for and download upstream releases, but we are upstream.

usaleem-ix replied:
Removed.

freqlabs commented on contrib/debian/openzfs-zfsutils.install:
```diff
@@ -0,0 +1,129 @@
+../tree/zfsutils-linux/* /
```
Ah this is where it was being used. We should be able to get rid of it.

usaleem-ix replied:
Removed.

usaleem-ix commented:
@cdluminate I have rebased and updated the bits to work with current master. Copyright file is trimmed to bare minimum and names of the past contributors for Debian packaging are preserved. `debian` directory is placed under `contrib/`.

I think we are ready for another round of review.

cdluminate commented on contrib/debian/README.Debian:
```diff
@@ -0,0 +1,68 @@
+zfs-linux for Debian
```
Could you please delete this file?

usaleem-ix replied:
Deleted.

cdluminate commented on contrib/debian/control:
```diff
@@ -0,0 +1,325 @@
+Section: contrib/kernel
+Priority: optional
+Maintainer: Debian ZFS on Linux maintainers <pkg-zfsonlinux-devel@alioth-lists.debian.net>
+Uploaders: Aron Xu <aron@debian.org>,
+           Carlos Alberto Lopez Perez <clopez@igalia.com>,
+           Mo Zhou <lumin@debian.org>
```
Please modify the uploader field to the correct maintainer address.

cdluminate replied:
Better to remove the whole `Uploaders:` field.

usaleem-ix replied:
Removed the `Uploaders:` field.

cdluminate commented on contrib/debian/control:
```diff
@@ -0,0 +1,325 @@
+Depends: ${misc:Depends}, ${shlibs:Depends}
+Breaks: libnvpair1, libnvpair3
+Replaces: libnvpair1, libnvpair3
+Conflicts: libnvpair3linux
```
Conflicts+replaces is better. I've mentioned this before. It's ok if you insist on not changing this anymore. Leaving it as is will likely cause problem when going back to Debian's official packaging.

usaleem-ix replied:
Sorry, `Replaces` was missed previously by mistake. I have added the package name in `Replaces`.

cdluminate commented on contrib/debian/gbp.conf:
```diff
@@ -0,0 +1,2 @@
+[DEFAULT]
```
Please delete this useless file. This is only useful for Debian official packaging. It is junk for a fork.

usaleem-ix replied:
Deleted.

cdluminate commented on contrib/debian/source/lintian-overrides:
```diff
@@ -0,0 +1 @@
+source: excessive-debhelper-overrides
```
remove file.

cdluminate commented:
It seems that I forgot to submit my first round if review. Please simply ignore my duplicated comments.

cdluminate commented on contrib/debian/changelog:
```diff
@@ -0,0 +1,1110 @@
+openzfs-linux (2.1.99-1) unstable; urgency=low
+
+  * Integrate minimally modified Debian packaging from ZFS on Linux
```
It is also very important to mention that this is a fork of `2.1.4-1` packaging. I will be unable to compare these files in the future if we don't have a version.

usaleem-ix replied:
I have mentioned the source `https://salsa.debian.org/zfsonlinux-team/zfs` and that this is a fork of `2.1.4-1` packaging.

cdluminate commented on contrib/debian/control:
```diff
@@ -0,0 +1,325 @@
+Source: openzfs-linux
+Section: contrib/kernel
+Priority: optional
+Maintainer: Debian ZFS on Linux maintainers <pkg-zfsonlinux-devel@alioth-lists.debian.net>
```
The correct value change is `Maintainer: iXsystems engineering team <dev@ixsystems.com>`.

usaleem-ix replied:
I have updated the address to OpenZFS mailing list address. Ideally anyone who wants to distribute packages built this way would change it to their address.

cdluminate commented:
A better way to simplify the copyright file is use the follow plain text:

```
This Debian packaging is a derived work of Debian's zfs-linux package [1].
The original copy of the Debian-styled copyright file can be found at [2].
The respective contributor information can be found in [2][3].

[1] https://tracker.debian.org/pkg/zfs-linux
[2] https://salsa.debian.org/zfsonlinux-team/zfs/-/blob/master/debian/copyright
[3] https://salsa.debian.org/zfsonlinux-team/zfs/-/blob/master/debian/changelog
```

7-line plain text is enough here to simplify this PR.

usaleem-ix commented on contrib/debian/README.Debian:
```diff
@@ -0,0 +1,68 @@
+
+ By default scrub jobs are scheduled on every second Sunday of month.
+
+ [ Periodic Trimming ]
```
Removed.

usaleem-ix commented on contrib/debian/NEWS:
```diff
@@ -0,0 +1,52 @@
+zfs-linux (0.8.3-1) unstable; urgency=medium
```
Removed.

usaleem-ix commented on contrib/debian/control:
```diff
@@ -0,0 +1,325 @@
+Uploaders: Aron Xu <aron@debian.org>,
+           Carlos Alberto Lopez Perez <clopez@igalia.com>,
+           Mo Zhou <lumin@debian.org>
+Build-Depends: abigail-tools (>= 1.8),
```
version dependency is removed.

usaleem-ix commented on contrib/debian/openzfs-zfs-dkms.lintian-overrides:
```diff
@@ -0,0 +1,6 @@
+binary-package-depends-on-toolchain-package Suggests: debhelper
```
All lintian overrides are removed.

usaleem-ix commented on contrib/debian/changelog:
```diff
@@ -0,0 +1,1110 @@
+
+ -- iXsystems engineering team <dev@ixsystems.com>  Fri, 30 Sep 2022 15:00:00 -0400
+
+zfs-linux (2.1.4-1) unstable; urgency=medium
```
There is only one changelog entry now. Rest of them are removed.

usaleem-ix commented on contrib/debian/copyright:
```diff
@@ -0,0 +1,400 @@
+ National Security, LLC.  The views and opinions of authors expressed
+ herein do not necessarily state or reflect those of the Untied States
+ Government or Lawrence Livermore National Security, LLC, and shall
+ not be used for advertising or product endorsement purposes.
```
Removed.

usaleem-ix commented on contrib/debian/copyright:
```diff
@@ -0,0 +1,400 @@
+ .
+ 1. Definitions.
+ .
+ 1.1. “Contributor” means each individual or entity that creates or contributes
```
Deleted.

usaleem-ix replied:
copyright file is updated as you suggested below in this [comment](https://github.com/openzfs/zfs/pull/13451#issuecomment-1270685687).

usaleem-ix commented on contrib/debian/openzfs-pyzfs-doc.lintian-overrides:
```diff
@@ -0,0 +1 @@
+possible-documentation-but-no-doc-base-registration
```
Removed.

usaleem-ix commented on contrib/debian/openzfs-zfs-test.lintian-overrides:
```diff
@@ -0,0 +1,11 @@
+arch-dependent-file-in-usr-share
```
Deleted.

usaleem-ix commented on contrib/debian/rules:
```diff
@@ -0,0 +1,282 @@
+override_dh_auto_test:
+ifeq (amd64,$(DEB_HOST_ARCH))
+	# Upstream provides an ABI guarantee that we validate here
+	-$(MAKE) checkabi
```
Removed the `-` sign.

freqlabs commented on contrib/debian/rules:
```diff
@@ -0,0 +1,281 @@
+override_dh_makeshlibs:
+	dh_makeshlibs -a -V
+ifeq ($(BUILD_UDEB), true)
+	dh_makeshlibs -V -plibnvpair3linux --add-udeb=libnvpair3-openzfs-udeb
+	dh_makeshlibs -V -plibuutil3linux --add-udeb=libuutil3-openzfs-udeb
+	dh_makeshlibs -V -plibzfs4linux --add-udeb=libopenzfs4-udeb
+	dh_makeshlibs -V -plibzfsbootenv1linux --add-udeb=libopenzfsbootenv1-udeb
+	dh_makeshlibs -V -plibzpool5linux --add-udeb=libopenzpool5-udeb
+	dh_makeshlibs -V -pzfsutils-linux --add-udeb=openzfsutils-udeb
+endif
+
+override_dh_strip:
+	dh_strip
+ifeq ($(BUILD_UDEB), true)
+	dh_strip -plibnvpair3-openzfs-udeb
+	dh_strip -plibuutil3-openzfs-udeb
+	dh_strip -plibopenzfs4-udeb
+	dh_strip -plibopenzfsbootenv1-udeb
+	dh_strip -plibopenzpool5-udeb
+	dh_strip -popenzfsutils-udeb
```
These names are all over the place. Did we settle on a convention? I think `openzfs-zfsutils` and `openzfs-libzpool5` would work a lot better, for example.

usaleem-ix replied:
Looks like I missed this part. Thanks for pointing out. I will update it.

usaleem-ix replied:
I have updated with new names here.

freqlabs commented on contrib/debian/rules:
```diff
@@ -0,0 +1,281 @@
+	dh_strip -popenzfs-libnvpair3-udeb
+	dh_strip -popenzfs-libuutil3-udeb
+	dh_strip -popenzfs-libzfs4-udeb
+	dh_strip -plopenzfs-libzfsbootenv1-udeb
```
Typo?

usaleem-ix replied:
Updated.

behlendorf commented on scripts/debian-packaging.sh:
```diff
@@ -0,0 +1,36 @@
+
+#
+# This script can be used to invoke OpenZFS build from native Debian
+# packaging.
```
I think it would be desirable to switch the `make deb` target to using the native packaging.  This way it's be regularly exercised by the CI and as part of the PR process.  Since we'll probably need to make some small CI tweaks to handle this I'm fine leaving it for a followup PR.

usaleem-ix replied:
Sure, I'll create a follow up PR for this.

behlendorf reviewed (commented):
It looks like all of the outstanding feedback has been addressed.  Is this PR ready to go, or is there still some outstanding work?

usaleem-ix commented:
> It looks like all of the outstanding feedback has been addressed. Is this PR ready to go, or is there still some outstanding work?

Yes all the feedback has been addressed. I think it's ready to go. It'd be great if @cdluminate could take another look if possible.

cdluminate commented on contrib/debian/control.udeb.in:
```diff
@@ -0,0 +1,86 @@
+Package: openzfs-libnvpair3-udeb
```
`*-udeb` packages are used for debian installer. I think you don't need all of them. Even for debian, these packages are not really enabled.

usaleem-ix replied:
@cdluminate I have removed all udeb sections and files.

cdluminate commented on contrib/debian/rules:
```diff
@@ -0,0 +1,281 @@
+		--with-linux-obj=$(KOBJ)
+	touch override_dh_configure_modules_stamp
+
+override_dh_binary-modules-udeb: override_dh_prep-deb-files override_dh_configure_modules_udeb
+	dh_testdir
+	dh_testroot
+	dh_prep
+
+	$(MAKE) $(NJOBS) -C $(CURDIR)/module modules
+
+	dh_installdirs -p${pmodules}-di
+	dh_install -p${pmodules}-di
+	dh_gencontrol -p${pmodules}-di
+
+	dh_builddeb -p${pmodules}-di
```
another `udeb` section.

usaleem-ix replied:
Removed.

usaleem-ix commented on contrib/debian/openzfs-zfsutils-udeb.install:
```diff
@@ -0,0 +1,2 @@
+../tmp/lib/udev/*_id    /sbin
```
Removed.

usaleem-ix commented on contrib/debian/rules:
```diff
@@ -0,0 +1,281 @@
+DKMSFILES := module include config zfs.release.in autogen.sh META AUTHORS \
+		COPYRIGHT LICENSE README.md
+
+ifndef BUILD_UDEB
```
Removed.

usaleem-ix commented:
> Apart from whether the packaging is synchronized with the latest packaging, it should be ok as long as it passes your build test after removing the last unnecessary bits.

@cdluminate Yes, I also ported the updates from `debian/2.1.6-2` release to keep it synchronized with the latest packaging bits for now. It is mentioned in changelog file and commit log (2.1.6-2 in place of 2.1.4-1).

cdluminate commented:
LGTM.

usaleem-ix commented:
> Looks ready to go apart from one other comment and a small nit in the commit message:
> 
> > All credits go to Debian ZFS on Linux team Packaging Team.
> 
> "team Packaging Team" ?

Fixed the commit message.

freqlabs reviewed (changes requested):
I realized I had not tested the -c option of the debian-packaging.sh script. It does not work correctly.

freqlabs commented on contrib/debian/rules:
```diff
@@ -0,0 +1,224 @@
+	dh_strip
+
+override_dh_auto_clean:
+	find . -name .gitignore -delete
```
All the files that get leftover are ignored by .gitignore, and `make clean` leaves a bunch of junk behind that is ignored by .gitignore as well, so maybe this is not a problem with the script after all, other than deleting the .gitignore files in the first place.

cdluminate replied:
This line of code is inherited from the salsa version of packaging, which was written in 2016 for some old zfs version. It is removable here.

usaleem-ix replied:
Thanks, I have removed this line. The leftover files should have been ignored by `.gitignore` but since it was removed, they were showing up. The leftover files are now ignored by `.gitignore`.

freqlabs reviewed (approved):
I'm happy enough with that.

behlendorf commented:
Merged, thanks everyone for tackling this!

dalbani commented:
I'm obviously commenting on a long closed ticket, but has this new capability been documented somewhere?
I've personally discovered it by browsing through the Git repo.
If I look at https://openzfs.github.io/openzfs-docs/Developer%20Resources/Building%20ZFS.html#configure-and-build, it says:
```
make deb #example for Debian/Ubuntu
```
I suppose it should also mention that `make native-deb` is available (even preferable?).
Unless I've missed something, I couldn't find a related commit in https://github.com/openzfs/openzfs-docs, as it was suggested in https://github.com/openzfs/zfs/pull/13451#issuecomment-1123956083.

usaleem-ix commented:
Thanks for the reminder @dalbani , I will create a PR for documenting this.
